AdSecGHTests/Parameters/AdSecSectionTests.cs

```diff
@@ -19,64 +19,59 @@
-      ISection actualSection = jsonParser.Sections[0];
+      var jsonParser = JsonParser.Deserialize(json);
+      var actualSection = jsonParser.Sections[0];
 
-      var expectedProfile = (IPerimeterProfile)flattened.Profile;
-      var actualProfile = (IPerimeterProfile)actualSection.Profile;
-
-      TestSection(flattened, actualSection);
+      TestSection(flattened, actualSection, true);
     }
 
     [Fact]
     public void SerialiaseFlattenedSectionTest() {
-      ISection section = CreateSection();
+      var section = CreateSection();
 
-      IDesignCode designCode = EN1992.Part1_1.Edition_2004.NationalAnnex.DE.Edition_2013;
+      var designCode = EN1992.Part1_1.Edition_2004.NationalAnnex.DE.Edition_2013;
       var adSec = IAdSec.Create(designCode);
-      ISection flattened = adSec.Flatten(section);
+      var flattened = adSec.Flatten(section);
 
       string fileName = Path.GetTempPath() + "AdSecSectionTest.ads";
       File.WriteAllText(fileName, CreateJson(designCode, flattened));
 
       string json = File.ReadAllText(fileName);
-      ParsedResult jsonParser = JsonParser.Deserialize(json);
-      ISection actualSection = jsonParser.Sections[0];
-
-      var expectedProfile = (IPerimeterProfile)flattened.Profile;
-      var actualProfile = (IPerimeterProfile)actualSection.Profile;
+      var jsonParser = JsonParser.Deserialize(json);
+      var actualSection = jsonParser.Sections[0];
 
       TestSection(flattened, actualSection);
     }
 
-    private void TestSection(ISection expected, ISection actual) {
+    private static void TestSection(ISection expected, ISection actual, bool unflattened = false) {
       var expectedProfile = (IPerimeterProfile)expected.Profile;
       var actualProfile = (IPerimeterProfile)actual.Profile;
 
       for (int i = 0; i < expectedProfile.SolidPolygon.Points.Count; i++) {
-        IPoint expectedPoint = expectedProfile.SolidPolygon.Points[i];
-        IPoint actualPoint = actualProfile.SolidPolygon.Points[i];
+        var expectedPoint = expectedProfile.SolidPolygon.Points[i];
+        var actualPoint = actualProfile.SolidPolygon.Points[i];
 
         Assert.Equal(expectedPoint.Y.Value, actualPoint.Y.Value, 4);
         Assert.Equal(expectedPoint.Z.Value, actualPoint.Z.Value, 4);
       }
-
-      Assert.Equal(((ISingleBars)expected.ReinforcementGroups[0]).Positions[0].Y.Value, ((ISingleBars)actual.ReinforcementGroups[0]).Positions[0].Y.Value, 4);
-      Assert.Equal(((ISingleBars)expected.ReinforcementGroups[0]).Positions[0].Z.Value, ((ISingleBars)actual.ReinforcementGroups[0]).Positions[0].Z.Value, 4);
-      Assert.Equal(((ISingleBars)expected.ReinforcementGroups[0]).BarBundle.CountPerBundle, ((ISingleBars)actual.ReinforcementGroups[0]).BarBundle.CountPerBundle);
-      Assert.Equal(((ISingleBars)expected.ReinforcementGroups[0]).BarBundle.Diameter, ((ISingleBars)actual.ReinforcementGroups[0]).BarBundle.Diameter);
+      if (!unflattened) {
```

Comment:
@SandeepArup is there a method for testing if a Section if unflattened or flatten? I don't like that we dictate if that is the case. There must be an internal flag we can reuse?

Reply:
@RaviKumarGubbala @vilayatk is there any flag which tell section is unflattened or flatten?

Reply:
Such kind of check shouldn't be there. Is it added becuase test are failing? I think the test fails becuase CountPerBundle cant be checked before after flatten. If there are three bars per bundle, it might be geting converted to three single bars. After confirming the behaviour we could update/delete the check related CountPerBundle. @SandeepArup

Reply:
I agree @RaviKumarGubbala. However, "It might be" OR always converting bar in bundle to single bars? 

These tests are json sterilization and deserialization test i.e. b/w flatten and original section and vice-versa and have only one bar in bundle. So, test should not fail in either case. @DominikaLos  lets discuss this tomorrow.

Reply:
@psarras I had some discussion with @RaviKumarGubbala and have feeling that after flattening section, profile will get changed(although section geometry will remain same) and so center of gravity and rebar position will be changed relative to new Origin/CG position and there is no guarantee that we can match the position with original section.

And second things, why these tests? Should we write test for third party library? If consuming Revit API or Grasshopper API, should we write test for their functionalities? If we are finding some issue in AdSec API then It should be reported to AdSec team for resolution.

So, in my opinion, these test should not be here.

Reply:
Hey! I agree we shouldn't be testing GSA.API here. I will need to look into more detail to understand if we simply remove them, however, I have the feeling they are also testing some secondary functionality, of which we need a test. So perhaps we morph this problem to something that helps us.